incubator/hnc/internal/reconcilers/object_test.go

```diff
@@ -15,6 +15,201 @@ import (
+	ctx := context.Background()
+
+	BeforeEach(func() {
+		// We want to ensure we're working with a clean slate, in case a previous tests objects still exist
+		cleanupObjects(ctx)
+	})
+
+	AfterEach(func() {
+		// Change current config back to the default value.
+		resetHNCConfigToDefault(ctx)
+		cleanupObjects(ctx)
+	})
+
+	Context("Add exception annotations ", func() {
+		const (
+			p  = "parent"
+			c1 = "child1"
+			c2 = "child2"
+			c3 = "child3"
+		)
+		tests := []struct {
+			name         string
+			selector     string
+			treeSelector string
+			want         []string
+			notWant      []string
+		}{
+			{name: "simply using select",
+				selector:     "!" + c1 + api.LabelTreeDepthSuffix,
+				treeSelector: "",
+				want:         []string{c2, c3},
+				notWant:      []string{c1},
+			},
+			{name: "using multiple select",
+				selector:     "!" + c1 + api.LabelTreeDepthSuffix + ", !" + c2 + api.LabelTreeDepthSuffix,
+				treeSelector: "",
+				want:         []string{c3},
+				notWant:      []string{c1, c2},
+			},
+			{name: "simply using treeSelect",
+				selector:     "",
+				treeSelector: "!" + c1,
+				want:         []string{c2, c3},
+				notWant:      []string{c1},
+			},
+			{name: "using multiple treeSelect",
+				selector:     "",
+				treeSelector: "!" + c1 + ", !" + c2,
+				want:         []string{c3},
+				notWant:      []string{c1, c2},
+			},
+			{name: "using both select and treeSelect negatively",
+				selector:     "!" + c1 + api.LabelTreeDepthSuffix,
+				treeSelector: "!" + c2,
+				want:         []string{c3},
+				notWant:      []string{c1, c2},
+			},
+			{name: "using both select and treeSelect positively",
+				selector:     c1 + api.LabelTreeDepthSuffix,
+				treeSelector: c2,
+				want:         []string{},
+				notWant:      []string{c1, c2, c3},
+			},
+		}
+
+		for _, tc := range tests {
+			// Making a local copy of tc is necessary to ensure the correct value is passed to the closure,
+			// for more details look at https://onsi.github.io/ginkgo/ and search for 'closure'
+			tc := tc
+			It("Should handle exceptions: "+tc.name, func() {
+				// Set up namespaces
+				fooName := createNS(ctx, "foo")
+				barName := createNS(ctx, "bar")
+				bazName := createNS(ctx, "baz")
+				quxName := createNS(ctx, "qux")
+
+				setParent(ctx, barName, fooName)
+				setParent(ctx, bazName, fooName)
+				setParent(ctx, quxName, fooName)
+
+				// Convert namespace index to real names
+				indToNm := map[string]string{
+					p:  fooName,
+					c1: barName,
+					c2: bazName,
+					c3: quxName,
+				}
+				tc.selector = replaceStrings(tc.selector, indToNm)
+				tc.treeSelector = replaceStrings(tc.treeSelector, indToNm)
+
+				// Create a Role with the selector and treeSelector annotation
+				a := make(map[string]string)
+				if tc.selector != "" {
+					a["propagate.hnc.x-k8s.io/select"] = tc.selector
+				}
+				if tc.treeSelector != "" {
+					a["propagate.hnc.x-k8s.io/treeSelect"] = tc.treeSelector
+				}
```

Comment:
What would happen if we made a selector with an empty string? Would that be "match everything" or "match nothing?" Do we define that properly in the design doc? If we treat an empty string as "missing" (i.e. match everything) then we could say:
```
makeObjectWithAnnotation(ctx, "Role", fooName, "testrole", map[string]string{
  "propagate.hnc.x-k8s.io/select": tc.selector,
  "propagate.hnc.x-k8s.io/treeSelect": tc.treeSelector,
})
```

which seems kind of nice.

Reply:
I tested it and seems like when converting string to selector, we do treat empty string as missing. Updated

Reply:
sgtm and can you please update the design doc to state that this is the intended behaviour?